src/symbolic_memory.ml

```diff
@@ -111,51 +133,81 @@ module M = struct
     let v0 = load_byte m a in
     loop a n 1 v0
 
+  (* TODO: *)
+  (* 1. Let pointers have symbolic offsets *)
+  (* 2. Let addresses have symbolic values *)
+  let calculate_address m (a : int32) : (Int32.t, Trap.t) Stdlib.Result.t =
+    match a.e with
+    | Val (Num (I32 i)) -> Ok i
+    | Ptr (base, offset) -> (
+      match Hashtbl.find_opt m.chunks base with
+      | None -> Error Trap.Memory_leak_use_after_free
+      | Some size ->
+        let ptr = Int32.add base (i32 offset) in
+        if ptr < base || ptr > Int32.add base (i32 size) then
+          Error Trap.Memory_heap_buffer_overflow
+        else Ok ptr )
+    | _ -> Format.kasprintf failwith "Cannot calculate address of: %a" Expr.pp a
+
   let load_8_s m a =
-    let v = loadn m (concretize_i32 a.e) 1 in
+    let* a = calculate_address m a in
+    let v = loadn m a 1 in
     match v.e with
-    | Val (Num (I32 i8)) -> Value.const_i32 (Int32.extend_s 8 i8)
-    | _ -> Cvtop (ExtS 24, v) @: Ty_bitv S32
+    | Val (Num (I8 i8)) ->
+      Ok (Value.const_i32 (Int32.extend_s 8 (Int32.of_int i8)))
+    | _ -> Ok (Cvtop (ExtS 24, v) @: Ty_bitv S32)
```

Comment:
If you replace the `let*` by `let+` in the beginning, then you can remove the `Ok` in each case of the `match`.